### ISSUE TITLE
"Currently played note" highlight improvement

### ISSUE THREAD

**Issue by guiv42** [helge17/tuxguitar#365]
### Discussed in https://github.com/helge17/tuxguitar/discussions/364

<div type='discussions-op-text'>

<sup>Originally posted by **logari84** April 26, 2024</sup>
Hi,

Tux guitar is an amazing application, though I think that the highlight of the currently played note could be improved. Sometimes it is really hard to see the highlighted not, I guess for people who are color-blinded it would be even harder. A better approach would be to place the currently played note inside a box (with curved corners it would look really nice) and with a pale background color. Apparently the box would contain both the tablature and the score.

I hope you find my feedback useful!

Thank you.</div>

**Comment by guiv42** [comment]
Just for information: a similar enhancement was suggested in 2021 on [sourceforge](https://sourceforge.net/p/tuxguitar/feature-requests/90/)

**Comment by guiv42** [comment]
I'm currently looking at a possible implementation, here is what it could look like (with configurable colors)

https://github.com/helge17/tuxguitar/assets/129443524/a3ffb940-dfb9-4a4c-85cc-65c1f4a07651

**Comment by guiv42** [comment]
@logari84 or anyone: I would appreciate some feedback on this proposal.
For sure it makes currently played beat much more visible, as required. But I fear it might be a bit visually aggressive, especially for fast parts with short beats duration.
There is always a possibility to make this configurable (i.e. rectangle around beat / around measure), but I would prefer to avoid this if possible.

**Comment by guiv42** [comment]
I finally made it configurable, see menu "View / Highlight Played Beat"
By default:
- in a fresh install of TuxGuitar, rectangle around played beat **is** displayed (as video above)
- for users upgrading from a previous version, rectangle around played beat **is not** displayed

This is a consequence of an implementation constraint

**Comment by helge17** [comment]
The highlighting looks very good! Perhaps a very bright red (e.g. 255,242,242) would go a little better with the red frame than the green (225,255,225), but that is really a matter of taste.

**Comment by guiv42** [comment]
Thanks for the feedback, and also for the menu icon it's perfect.
About the color, feel free to change it if you think it looks better. I went to a light green on a logical basis: considering played note was red, selection was blue, and all 3 colors could be displayed simultaneously. Honestly, I did not really consider any aesthetic criterion!
What do you think of the "location" of the menu item / option in toolbar? When looking at the existing interface it seemed a logical place, but after re-reading the [view section](https://github.com/helge17/tuxguitar/blob/87942f990093f83b798e9638e68a6ea1ad46cdf6/desktop/TuxGuitar/share/help/detail_view.html#L73) of user doc I'm not convinced it's the best place. Or should we add an horizontal separator?

**Comment by helge17** [comment]
Since the setting affects the playback of the song, I would perhaps add it at the end of the "Player" menu, including a separator.
One small point: would it be possible to keep the highlighting when playback is paused?

**Comment by guiv42** [comment]
>  I would perhaps add it at the end of the "Player" menu, including a separator.

Why not, I really don't know what's best. Do you want me to move it there?

> One small point: would it be possible to keep the highlighting when playback is paused?

I don't think this is easily feasible. Practically it seems to me there is no such thing as a "paused" state for the player. Whenever user clicks *stop* or *pause*, the player *stops*. The only difference between stop and pause is where the caret goes to.
If the last played beat remains highlighted after pausing, on which conditions should it become normal again?

By the way that really puzzles me: there *is* some code explicitly calling `MidiPlayer.isPaused()`, for example in [TGTransportDialog](https://github.com/helge17/tuxguitar/blob/87942f990093f83b798e9638e68a6ea1ad46cdf6/desktop/TuxGuitar/src/org/herac/tuxguitar/app/view/dialog/transport/TGTransportDialog.java#L310) (there are other places with similar code). The only difference between `paused` and `not paused` is to select different source files for the buttons icons. Extract from TGTransportDialog:
```
if(getStatus() == STATUS_PAUSED){
    this.first.setImage(TuxGuitar.getInstance().getIconManager().getTransportFirst2());
    (...)
}else if(getStatus() == STATUS_STOPPED){
        this.first.setImage(TuxGuitar.getInstance().getIconManager().getTransportFirst1());
```

And I really **don't understand** this, as it seems the 2 icons are exactly identical!
See for example Oxygen: [transportFirst1](https://github.com/helge17/tuxguitar/blob/master/desktop/TuxGuitar/share/skins/Oxygen/transport_first_1.png) and [transportFirst2](https://github.com/helge17/tuxguitar/blob/master/desktop/TuxGuitar/share/skins/Oxygen/transport_first_2.png), these 2 files *are* exactly identical.

**Comment by helge17** [comment]
>>    I would perhaps add it at the end of the "Player" menu, including a separator.

> Why not, I really don't know what's best. Do you want me to move it there?

I will move it there, if that is okay with you, so that you can take care of more complex tasks that I am not able to do.

> ... it seems the 2 icons are exactly identical!

The highlighting in the pause status is just an idea, nothing important. But I'll take a look at the two icons.

**Comment by guiv42** [comment]
about the icons: surprisingly, there *are* some differences in transport-related icons *for some skins*. I checked: transport_first_x, transport_last_x, transport_previous_x, transport_next_x, transport_stop_x, transport_play_x,
transport_icon_first_x, transport_icon_last_x, transport_icon_previous_x, transport_icon_next_x, transport_icon_stop_x, transport_icon_play_x
- blue_serious: all icons with suffixes _1 and _2 differ
- ersplus: all icons with suffixes _1 and _2 differ
- Lavender: transport_icon_last_1 and _2 files *are* different, even if visually I don't see the difference, all other icons _1 and _2 are identical
- all other skins: all icons _1 and _2 are identical

I don't understand the logic behind that. Even looking at the historical behavior of the player. See this video, made with TuxGuitar 1.5.6, clicking on "stop" changes player buttons icons. But why?

https://github.com/helge17/tuxguitar/assets/129443524/2a155d43-be15-4578-9d2d-4973ede54fa7

If that's OK for you I can create a new issue for this, some cleanup might be useful.

**Comment by helge17** [comment]
The menu item "Highlight Played Beat" is now at the bottom of the "Player" main menu (b7368c2f362d398ef148ae3eea712c95d8a839ab). I also changed the background of the played measure from green to red (b0b9d9f77f81ead3a353765d2ff85980db882456), I hope that's ok. It seems more consistent to me to have the same color for highlighting the measure and the beat.

It would be great if you could open a new issue for the transport-related icons, I'll look into it.

**Comment by helge17** [comment]
I have noticed a small issue with the frame around the played notes: If you deactivate the tablature and only display the score, the frame does not include the notes below the bottom line:

![grafik](https://github.com/helge17/tuxguitar/assets/5897631/11afd887-9c5e-46fe-8ee7-08eb4495d6ab)

**Comment by guiv42** [comment]
Thanks for the menu and background color

> I have noticed a small issue with the frame around the played notes

OK I'll fix that

**Comment by guiv42** [comment]
That should be OK now (#378)


https://github.com/helge17/tuxguitar/assets/129443524/56c1648f-ef11-4e6f-9bc2-d0c4d97d43ee



I also fixed another issue: clicking on "Default" button in "Tools/Settings" dialog did not restore the default value for "played measure background color" parameter

**Comment by guiv42** [comment]
Issue about player icons is now created.
No intention to go any further on my side about this issue here. Can we close it now?

**Comment by helge17** [comment]
> Can we close it now?

I think so.

**Comment by guiv42** [comment]
Now that app version is available in user config file (#376), I'll try to ensure that the new parameter is True by default, so that it doesn't become a "hidden" feature for users upgrading from 1.6.x.

**Comment by guiv42** [comment]
done
